Mathlib/NumberTheory/Padics/PadicIntegers.lean

```diff
@@ -0,0 +1,703 @@
+
+@[simp]
+theorem valuation_one : valuation (1 : ℤ_[p]) = 0 := Padic.valuation_one
+#align padic_int.valuation_one PadicInt.valuation_one
+
+@[simp]
+theorem valuation_p : valuation (p : ℤ_[p]) = 1 := by simp [valuation]
+#align padic_int.valuation_p PadicInt.valuation_p
+
+theorem valuation_nonneg (x : ℤ_[p]) : 0 ≤ x.valuation := by
+  by_cases hx : x = 0
+  · simp [hx]
+  have h : (1 : ℝ) < p := by exact_mod_cast hp.1.one_lt
+  rw [← neg_nonpos, ← (zpow_strictMono h).le_iff_le]
+  show (p : ℝ) ^ (-valuation x) ≤ (p : ℝ) ^ (0 : ℤ)
+  rw [← norm_eq_pow_val hx]
+  simpa using x.property
+#align padic_int.valuation_nonneg PadicInt.valuation_nonneg
+
+@[simp]
+theorem valuation_p_pow_mul (n : ℕ) (c : ℤ_[p]) (hc : c ≠ 0) :
+    ((p : ℤ_[p]) ^ n * c).valuation = n + c.valuation := by
+  have : ‖(p : ℤ_[p]) ^ n * c‖ = ‖(p : ℤ_[p]) ^ n‖ * ‖c‖ := norm_mul _ _
+  have aux : (p : ℤ_[p]) ^ n * c ≠ 0 := by
+    contrapose! hc
+    rw [mul_eq_zero] at hc
+    cases' hc with hc hc
+    · refine (hp.1.ne_zero ?_).elim
+      exact_mod_cast pow_eq_zero hc
+    · exact hc
+  rwa [norm_eq_pow_val aux, norm_p_pow, norm_eq_pow_val hc, ← zpow_add₀, ← neg_add,
+    zpow_inj, neg_inj] at this
+  · exact_mod_cast hp.1.pos
+  · exact_mod_cast hp.1.ne_one
+  · exact_mod_cast hp.1.ne_zero
+#align padic_int.valuation_p_pow_mul PadicInt.valuation_p_pow_mul
+
+section Units
+
+/-! ### Units of `ℤ_[p]` -/
+
+-- Porting note: `reducible` cannot be local and making it global breaks a lot of things
+-- attribute [local reducible] PadicInt
+
+theorem mul_inv : ∀ {z : ℤ_[p]}, ‖z‖ = 1 → z * z.inv = 1
+  | ⟨k, _⟩, h => by
+    have hk : k ≠ 0 := fun h' => zero_ne_one' ℚ_[p] (by simp [h'] at h)
+    unfold PadicInt.inv
+    rw [norm_eq_padic_norm] at h
+    dsimp only
+    rw [dif_pos h]
+    apply Subtype.ext_iff_val.2
+    simp [mul_inv_cancel hk]
+#align padic_int.mul_inv PadicInt.mul_inv
+
+theorem inv_mul {z : ℤ_[p]} (hz : ‖z‖ = 1) : z.inv * z = 1 := by rw [mul_comm, mul_inv hz]
+#align padic_int.inv_mul PadicInt.inv_mul
+
+theorem isUnit_iff {z : ℤ_[p]} : IsUnit z ↔ ‖z‖ = 1 :=
+  ⟨fun h => by
+    rcases isUnit_iff_dvd_one.1 h with ⟨w, eq⟩
+    refine' le_antisymm (norm_le_one _) _
+    have := mul_le_mul_of_nonneg_left (norm_le_one w) (norm_nonneg z)
+    rwa [mul_one, ← norm_mul, ← eq, norm_one] at this, fun h =>
+    ⟨⟨z, z.inv, mul_inv h, inv_mul h⟩, rfl⟩⟩
+#align padic_int.is_unit_iff PadicInt.isUnit_iff
+
+theorem norm_lt_one_add {z1 z2 : ℤ_[p]} (hz1 : ‖z1‖ < 1) (hz2 : ‖z2‖ < 1) : ‖z1 + z2‖ < 1 :=
+  lt_of_le_of_lt (nonarchimedean _ _) (max_lt hz1 hz2)
+#align padic_int.norm_lt_one_add PadicInt.norm_lt_one_add
+
+theorem norm_lt_one_mul {z1 z2 : ℤ_[p]} (hz2 : ‖z2‖ < 1) : ‖z1 * z2‖ < 1 :=
+  calc
+    ‖z1 * z2‖ = ‖z1‖ * ‖z2‖ := by simp
+    _ < 1 := mul_lt_one_of_nonneg_of_lt_one_right (norm_le_one _) (norm_nonneg _) hz2
+
+#align padic_int.norm_lt_one_mul PadicInt.norm_lt_one_mul
+
+@[simp, nolint simpNF] -- Porting note: simpNF wants to change the LHS to ¬ IsUnit z
```

Comment:
The linter is correct to flag this lemma. A better solution than switching off the linter should be found, maybe adding other lemmas, or maybe just removing `simp` from this lemma altogether.

Reply:
I guess the easiest solution is to remove the `simp` so I removed this one too ;)